### PR TITLE
[DataPipes] Add group support to the sharding_filter

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88424

Differential Revision: [D41006747](https://our.internmc.facebook.com/intern/diff/D41006747)

### CODE REVIEW

File: torch/utils/data/datapipes/iter/grouping.py [hunk 1]
```diff
@@ -25,17 +32,45 @@ class ShardingFilterIterDataPipe(IterDataPipe):
     Args:
         source_datapipe: Iterable DataPipe that will be sharded
     """
-    def __init__(self, source_datapipe: IterDataPipe):
+
+    def __init__(self, source_datapipe: IterDataPipe, sharding_group_filter=None):
         self.source_datapipe = source_datapipe
+        self.sharding_group_filter = sharding_group_filter
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/grouping.py:38]: Do we need an extra API to set `sharding_group_filter`?

Based on the implementation, it seems `sharding_group_filter` is an integer, could we change it `set` or `list` to support multiple filters?

File: torch/utils/data/datapipes/iter/grouping.py [hunk 2]
```diff
@@ -25,17 +32,45 @@ class ShardingFilterIterDataPipe(IterDataPipe):
...
+        self._update_num_of_instances()
+
+    def _update_num_of_instances(self):
+        sorted_sharding_groups = []
+        for key in sorted(self.groups.keys()):
+            if self.sharding_group_filter is None:
+                sorted_sharding_groups.append(self.groups[key])
+            else:
+                if key == self.sharding_group_filter:
+                    sorted_sharding_groups.append(self.groups[key])
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/grouping.py:64]: nit:
```py
if self.sharding_group_filter is None or key == self.sharding_group_filter:
    sorted_sharding_groups.append(self.groups[key])
```

File: torch/utils/data/datapipes/iter/grouping.py [hunk 3]
```diff
@@ -25,17 +32,45 @@ class ShardingFilterIterDataPipe(IterDataPipe):
...
         self.instance_id = 0
+        self._update_num_of_instances()
 
     def is_shardable(self):
         return True
 
-    def apply_sharding(self, num_of_instances, instance_id):
-        self.num_of_instances = num_of_instances
-        self.instance_id = instance_id
+    def apply_sharding(self, num_of_instances, instance_id, sharding_group=SHARDING_PRIORITIES.DEFAULT):
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/grouping.py:47]: Super nit: Could we add a validation that `instance_id < num_of_instances`?

File: test/test_datapipe.py [hunk 4]
```diff
@@ -2663,6 +2667,30 @@ def test_simple_sharding(self):
...
+            pipe.apply_sharding(3, 1, sharding_group=300)
+
+        actual = list(dp)
+        expected = [17, 47, 77]
+        self.assertEquals(expected, actual)
+        self.assertEquals(3, len(dp))
+
+        with self.assertRaises(Exception):
+            dp.apply_sharding(2, 1, sharding_group=SHARDING_PRIORITIES.DEFAULT)
+            dp.apply_sharding(5, 3, sharding_group=SHARDING_PRIORITIES.MULTIPROCESSING)
```
Comment (ejguan) [inline test/test_datapipe.py:2692]: Could we add a separate context of `self.assertRaises` for the second Error?

**Review by ejguan — APPROVED** [review]
LGTM with a few nit comments.

File: torch/utils/data/datapipes/iter/grouping.py (hunk 1 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/grouping.py:38]: I have no use-cases for it, but it will be trivial to change later if we require to.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41006747).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41006747).

File: torch/utils/data/graph_settings.py [hunk 5]
```diff
@@ -42,7 +46,7 @@ def apply_sharding(datapipe: DataPipe, num_of_instances: int, instance_id: int)
                     if already_applied_to is not None:
                         raise RuntimeError('This implementation of sharding can be only applied once per instance of DataPipeline.',
                                            'Already applied to', already_applied_to, 'while trying to apply to', pipe)
-                    pipe.apply_sharding(num_of_instances, instance_id)
+                    pipe.apply_sharding(num_of_instances, instance_id, sharding_group=sharding_group)
```
Comment (wenleix) [inline torch/utils/data/graph_settings.py:49]: noob question: does `is_shardable` and `apply_sharding` only exists in `ShardingFilterIterDataPipe` ? 

Also if there is no `ShardingFilterIterDataPipe`, look like no sharding will happen? (shall we error in that case? ) :) 

cc @ejguan

**Pull request closed without merging**